agent/consul/autopilotevents/ready_servers_events.go

```diff
@@ -0,0 +1,305 @@
+	snapshot     []stream.Event
+}
+
+func NewReadyServersEventPublisher(config Config) *ReadyServersEventPublisher {
+	return &ReadyServersEventPublisher{
+		Config: config,
+		snapshot: []stream.Event{
+			{
+				Topic:   EventTopicReadyServers,
+				Index:   0,
+				Payload: EventPayloadReadyServers{},
+			},
+		},
+	}
+}
+
+//go:generate mockery -name StateStore -inpkg -testonly
+type StateStore interface {
+	GetNodeID(types.NodeID, *acl.EnterpriseMeta) (uint64, *structs.Node, error)
+}
+
+//go:generate mockery -name Publisher -inpkg -testonly
+type Publisher interface {
+	Publish([]stream.Event)
+}
+
+//go:generate mockery -name timeProvider -inpkg -testonly
+type timeProvider interface {
+	Now() time.Time
+}
+
+// PublishReadyServersEvents will publish a "ready-servers" event if the list of
+// ready servers has changed since the last time events were published.
+func (r *ReadyServersEventPublisher) PublishReadyServersEvents(state *autopilot.State) {
+	if events, ok := r.readyServersEvents(state); ok {
+		// if the event publisher were to not be able to keep up with procesing events
+		// then its possible this blocks. It could cause autopilot to not update its
+		// state as often as it should. However if this blocks for over 10s then
+		// not updating the autopilot state as quickly is likely the least of our
+		// concerns. If we need to make this async then we probably need to single
+		// flight these to ensure proper event ordering.
+		r.snapshotLock.Lock()
+		r.snapshot = events
+		r.snapshotLock.Unlock()
+		r.Publisher.Publish(events)
+	}
+}
+
+func (r *ReadyServersEventPublisher) readyServersEvents(state *autopilot.State) ([]stream.Event, bool) {
+	// First, we need to pull all the ready servers out from the autopilot state.
+	servers := r.autopilotStateToReadyServers(state)
+
+	// Next we, sort the servers list to make comparison easier later on. We do
+	// this outside of the next length check conditional block to ensure that all
+	// values of previousReadyServers we store will be sorted and the future
+	// comparison's will remain valid.
+	sort.Slice(servers, func(i, j int) bool {
+		// no two servers can have the same id so this is sufficient
+		return servers[i].ID < servers[j].ID
+	})
+
+	// If the number of ready servers hasn't changed then we need to inspect individual
+	// servers to see if there are differences. If the number of servers has changed
+	// we know that an event should be generated and sent.
+	if len(r.previous) == len(servers) {
+		diff := false
+		// We are relying on the fact that both of the slices will be sorted and that
+		// we don't care what the actual differences are but instead just that they
+		// have differences. We can then just iterate through all elements of both
+		// slices at the same time and check the elements for equality. Also,
+		// this is taking advantage of Go being able to do equality checking on
+		// structs without having to check all the fields separately. If we
```

Comment:
Is this comment still correct? It looks like we're comparing the fields separately now.

Reply:
It is no longer accurate. I have removed that part of the comment.